src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -1006,64 +1010,112 @@ public SolutionCompilationState WithOptions(SolutionOptionSet options)
+            var newTracker = tracker.FreezePartialState(this, cancellationToken);
+
+            Contract.ThrowIfFalse(newIdToProjectStateMap.ContainsKey(projectId));
+            newIdToProjectStateMap = newIdToProjectStateMap.SetItem(projectId, newTracker.ProjectState);
+            newIdToTrackerMap = newIdToTrackerMap.SetItem(projectId, newTracker);
+        }
+
+        var newState = this.SolutionState.Branch(
+            idToProjectStateMap: newIdToProjectStateMap,
+            dependencyGraph: SolutionState.CreateDependencyGraph(this.SolutionState.ProjectIds, newIdToProjectStateMap));
+        var newCompilationState = this.Branch(
+            newState,
+            newIdToTrackerMap);
+
+        // Set the frozen solution to be its own frozen solution.  Freezing multiple times is a no-op.
+        newCompilationState._cachedFrozenSnapshot = _cachedFrozenSnapshot;
+
+        return newCompilationState;
+    }
+
     /// <summary>
-    /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time, assuming a background compiler is
-    /// busy building this compilations.
-    ///
-    /// A compilation for the project containing the specified document id will be guaranteed to exist with at least the syntax tree for the document.
-    ///
+    /// Creates a branch of the solution that has its compilations frozen in whatever state they are in at the time,
+    /// assuming a background compiler is busy building this compilations.
+    /// <para/>
+    /// A compilation for the project containing the specified document id will be guaranteed to exist with at least the
+    /// syntax tree for the document.
+    /// <para/>
     /// This not intended to be the public API, use Document.WithFrozenPartialSemantics() instead.
     /// </summary>
-    public SolutionCompilationState WithFrozenPartialCompilationIncludingSpecificDocument(
+    public AsyncLazy<SolutionCompilationState> WithFrozenPartialCompilationIncludingSpecificDocument(
         DocumentId documentId, CancellationToken cancellationToken)
     {
-        // in progress solutions are disabled for some testing
-        if (this.Services.GetService<IWorkspacePartialSolutionsTestHook>()?.IsPartialSolutionDisabled == true)
-            return this;
+        AsyncLazy<SolutionCompilationState>? lazyState;
+        using (_cachedFrozenDocumentStateLock.DisposableWait(cancellationToken))
+        {
+            if (!_cachedFrozenDocumentState.TryGetValue(documentId, out lazyState))
+            {
+                lazyState = new AsyncLazy<SolutionCompilationState>(
+                    cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: false, cancellationToken),
+                    cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: true, cancellationToken).GetAwaiter().GetResult());
```

Comment:
>cancellationToken => ComputeFrozenPartialCompilationIncludingSpecificDocumentAsync(documentId, synchronous: true, cancellationToken).GetAwaiter().GetResult() [](http://example.com/codeflow?start=20&length=157)

Learning time: What's the advantage of doing this over not sending in a synchronous function?

Reply:
updated this.  it's a but more code duplication, but makes it clear how one function is sync, and the other async.  teh benefit of hte original pattern is sharing as much code as possible.  However, it is a big uglier as you have to be very careful to do no async work when `synchronous: true` is passed in, and you then have to know it's not actually going to block the returned Task when .GetAwaiter().Result is used.